_posts/2024_07_07-EOF.md

```diff
@@ -0,0 +1,101 @@
+
+## Remove JUMPDEST analysis
+
+Probably the biggest advantage of EOF is that the JUMPDEST analysis is performed at deploy time instead of at runtime. The JUMPDEST analysis checks which parts of the byte code are data and which are valid jump destinations. This needs to be done, so that a JUMP opcode does not end up in the immediate of for example a PUSH32. 
+
+This JUMPDEST analysis can be cached in theory, but there have been attacks in the past where nodes could've been DOS'd by creating contracts with a lot of initcode over and over again. In order to combat these DOS issues, the maximum initcode size was introduced in EIP-3860. 
+
+In theory the JUMPDEST analysis limits the maximum size of a contract that can be deployed. In my opinion however, the max codes size can be increased significantly without running into these issues again. The DOS back then was because the initcode size was unlimited and EIP-3860 introduced additional gas costs for the JUMPDEST analysis.
+
+## Static program analysis
+
+As long as the EVM contains non-static jumps, the analysis of a program is in the worst case O(n^2). EOF removes dynamic jumps which reduces the analysis to O(n) in the worst case. 
+
+I'm not convinced that that the difference is significant though, since EVM programs are inherently small in size. Even at the worst case of 24576 byte, the analysis needs ~600 million steps. Assuming 1000ns per step (which is quite a lot, for reference KECCAK takes <700ns on my machine) this can be analysis be done in 10 minutes.
+
+## Smaller program size
+
+One advantage of EOF is smaller program size. The added structure of EOF allows compilers to optimize the way they're structuring their code which might reduce the program size by 2-5% according to Charles Cooper from Vyper.
+
+## Move runtime checks to deploy time
+
+EOF moves a lot of runtime checks to deploy time. For example the checks for stack overflow and stack underflow are done at deploy time. This would in theory make the execution faster. 
+
+There are two issues in practice that I see. We will still need to maintain the un-optimized opcodes for the legacy EVM, which means duplicating a lot of logic, once with the additional checks, once without. And the speed improvements for the average contracts are most likely miniscule since most of the latency of these checks is probably optimized away by the processor when loading memory. 
+
+I would really like to see the speed improvements of a mainnet block with and without these checks.
+
+## Address Space Extension
+
+EOF does not allow for EXTCALL, EXTDELEGATECALL or EXTSTATICCALL to call addresses where the upper 12 bytes are set. This restriction allows to set some of these in the future as a bitmask for address space extension.
+
+However this is not required in legacy contracts, which makes it impossible to move to address space extension for the legacy evm anyway.
```

Comment:
There was an interesting idea by axic a while ago to allow legacy contracts to communicate with 32-byte addresses. Not sure if worth mentioning: https://notes.ethereum.org/@axic/ase-bridge-contracts also https://notes.ethereum.org/@ipsilon/address-space-extension-exploration